storage-node-v2/src/services/sync/storageObligations.ts

```diff
@@ -0,0 +1,217 @@
+}
+
+/**
+ * Data object abstraction.
+ */
+type DataObject = {
+  /**
+   * Data object ID
+   */
+  id: string
+
+  /**
+   * Assigned bag ID
+   */
+  bagId: string
+}
+
+/**
+ * Get storage provider obligations like (assigned data objects) from the
+ * runtime (Query Node).
+ *
+ * @param queryNodeUrl - Query Node URL
+ * @param workerId - worker ID
+ * @returns promise for the DataObligations
+ */
+export async function getStorageObligationsFromRuntime(
+  queryNodeUrl: string,
+  workerId: number
+): Promise<DataObligations> {
+  const api = new QueryNodeApi(queryNodeUrl)
+
+  const allBuckets = await getAllBuckets(api)
+
+  const bucketIds = allBuckets
+    .filter((bucket) => bucket.operatorStatus?.workerId === workerId)
+    .map((bucket) => bucket.id)
+  const assignedBags = await getAllAssignedBags(api, bucketIds)
+
+  const bagIds = assignedBags.map((bag) => bag.id)
+  const assignedDataObjects = await getAllAssignedDataObjects(api, bagIds)
+
+  const model: DataObligations = {
+    storageBuckets: allBuckets.map((bucket) => ({
+      id: bucket.id,
+      operatorUrl: bucket.operatorMetadata?.nodeEndpoint ?? '',
+      workerId: bucket.operatorStatus?.workerId,
+    })),
+    bags: assignedBags.map((bag) => ({
+      id: bag.id,
+      buckets: bag.storageAssignments.map((bucketInBag) => bucketInBag.storageBucket.id),
+    })),
+    dataObjects: assignedDataObjects.map((dataObject) => ({
+      id: dataObject.id,
+      bagId: dataObject.storageBagId,
+    })),
+  }
+
+  return model
+}
+
+/**
+ * Get storage bucket IDs assigned to the worker.
+ *
+ * @param queryNodeUrl - Query Node URL
+ * @param workerId - worker ID
+ * @returns storage bucket IDs
+ */
+export async function getStorageBucketIdsByWorkerId(queryNodeUrl: string, workerId: number): Promise<string[]> {
+  const api = new QueryNodeApi(queryNodeUrl)
+
+  const idFragments = await api.getStorageBucketDetailsByWorkerId(workerId.toString())
+  const ids = idFragments.map((frag) => frag.id)
+
+  return ids
+}
+
+/**
+ * Get IDs of the data objects assigned to the bag ID.
+ *
+ * @param api - initialiazed QueryNodeApi instance
+ * @param bagId - bag ID
+ * @returns data object IDs
+ */
+export async function getDataObjectIDsByBagId(queryNodeUrl: string, bagId: string): Promise<string[]> {
+  const api = new QueryNodeApi(queryNodeUrl)
+  const dataObjects = await getAllAssignedDataObjects(api, [bagId])
+
+  return dataObjects.map((obj) => obj.id)
+}
+
+/**
+ * Get all storage buckets registered in the runtime (Query Node).
+ *
+ * @param api - initialiazed QueryNodeApi instance
+ * @returns storage buckets data
+ */
+async function getAllBuckets(api: QueryNodeApi): Promise<StorageBucketDetailsFragment[]> {
+  const idFragments = await api.getStorageBucketIds()
+  const ids = idFragments.map((frag) => frag.id)
+
+  return await getAllObjectsWithPaging(
+    'get all storage buckets',
+    async (offset, limit) => await api.getStorageBucketDetails(ids, offset, limit)
+  )
```

Comment:
Using `await api.getStorageBucketDetails(ids, offset, limit)` defeats the purpose of using `connection` query in the first place for retrieving the ids. I think this should be:
```suggestion
  const idFragments = await api.getStorageBucketIds()
  const ids = idFragments.map((frag) => frag.id)

  return await getAllObjectsWithPaging(
    'get all storage buckets',
    async (offset, limit) => await api.getStorageBucketDetails(ids.slice(offset, offset + limit))
  )
```
And the query itself should not specify `offset`

Reply:
Fixed.